### PR TITLE
Plataform+Flow: Add support to get/set hostname.

### PR DESCRIPTION
**Pull request by cabelitos**
Signed-off-by: Guilherme Iscaro <guilherme.iscaro@intel.com>

@barbieri Could you please take a look in this PR?

### CODE REVIEW

File: src/lib/common/include/sol-platform.h [hunk 1]
```diff
@@ -190,6 +190,53 @@ int sol_platform_get_mount_points(struct sol_ptr_vector *vector);
...
+ *
+ * If the hostname changes @c cb will be called.
+ *
+ * @param cb
+ * @param data
+ * @return 0 on success, negative errno otherwise.
+ *
+ * @see sol_plataform_del_hostname_monitor()
+ */
+int sol_plataform_add_hostname_monitor(void (*cb)(void *data), const void *data);
```
Comment (barbieri) [inline src/lib/common/include/sol-platform.h]: please specify the const char *hostname in the cb

File: src/lib/common/include/sol-platform.h [hunk 2]
```diff
@@ -190,6 +190,53 @@ int sol_platform_get_mount_points(struct sol_ptr_vector *vector);
...
 /**
+ * @brief Gets the hostname.
+ *
+ * @param name The current hostname.
+ *
+ * @return 0 on success, negative errno otherwise.
+ *
+ * @see sol_plataform_set_hostname()
+ */
+int sol_plataform_get_hostname(char **name);
```
Comment (barbieri) [inline src/lib/common/include/sol-platform.h]: I'd say we better return the `char *` or hide the memory management complexity and return `const char *`, then internally we can allocate or keep a fixed static buffer, it's easier to use.

File: src/lib/common/sol-platform-impl-contiki.c [hunk 3]
```diff
@@ -135,3 +135,31 @@ sol_platform_impl_umount(const char *mpoint, void (*cb)(void *data, const char *
     SOL_WRN("Not implemented");
     return -ENOTSUP;
 }
+
+int
+sol_plataform_unregister_hostname_monitor(void)
+{
+    SOL_WRN("Not implemented");
+    return -ENOTSUP;
```
Comment (barbieri) [inline src/lib/common/sol-platform-impl-contiki.c]: @ibriano do we have a hostname concept in contiki?

File: src/lib/common/sol-platform-impl-linux-micro.c [hunk 4]
```diff
@@ -1001,3 +1012,70 @@ sol_platform_linux_micro_inform_service_state(const char *service, enum sol_plat
...
     sol_platform_inform_service_monitors(service, state);
 }
+
+int
+sol_plataform_impl_set_hostname(const char *name)
+{
+    size_t len = strlen(name);
+
+    if (len > HOST_NAME_MAX) {
+        SOL_WRN("Hostname can not be bigger than %d", HOST_NAME_MAX);
```
Comment (barbieri) [inline src/lib/common/sol-platform-impl-linux-micro.c]: put the "%s" as well

File: src/lib/common/sol-platform-impl-linux-micro.c [hunk 5]
```diff
@@ -1001,3 +1012,70 @@ sol_platform_linux_micro_inform_service_state(const char *service, enum sol_plat
...
+    return true;
+}
+
+int
+sol_plataform_register_hostname_monitor(void)
+{
+    if (hostname_monitor.watcher)
+        return 0;
+
+    hostname_monitor.fd = open("/proc/sys/kernel/hostname", O_RDONLY);
```
Comment (barbieri) [inline src/lib/common/sol-platform-impl-linux-micro.c]: `O_CLOEXEC`

File: src/lib/common/sol-platform-impl-riot.c [hunk 6]
```diff
@@ -208,3 +208,31 @@ sol_platform_impl_umount(const char *mpoint, void (*cb)(void *data, const char *
     SOL_WRN("Not implemented");
     return -ENOTSUP;
 }
+
+int
+sol_plataform_unregister_hostname_monitor(void)
+{
+    SOL_WRN("Not implemented");
```
Comment (barbieri) [inline src/lib/common/sol-platform-impl-riot.c]: @ibriano do we have hostname in riot?

File: src/lib/common/sol-platform-impl-systemd.c [hunk 7]
```diff
@@ -489,3 +489,84 @@ sol_platform_impl_shutdown(void)
...
+sol_plataform_impl_get_hostname(char **name)
+{
+    sd_bus *bus;
+    sd_bus_error error = SD_BUS_ERROR_NULL;
+    int r;
+
+    bus = sol_bus_get(NULL);
+    SOL_NULL_CHECK(bus, -ENOTCONN);
+
+    r = sd_bus_get_property_string(bus, "org.freedesktop.hostname1",
```
Comment (barbieri) [inline src/lib/common/sol-platform-impl-systemd.c]: as I told you, it's simpler to just use `gethostname()` here, it's faster and you can't do it synchronously.

Usually set is the bitch, due permissions and so on. the daemon will do `sethosname()`, thus `gethostname()` will work

File: src/modules/flow/platform/platform.c [hunk 8]
```diff
@@ -208,4 +212,75 @@ platform_machine_id_open(struct sol_flow_node *node,
...
 
+static int
+hostname_send(struct sol_flow_node *node)
+{
+    int r;
+    char *name;
+
+    r = sol_plataform_get_hostname(&name);
+    SOL_INT_CHECK(r, < 0, r);
+    return sol_flow_send_string_packet(node, 0, name);
```
Comment (barbieri) [inline src/modules/flow/platform/platform.c]: see, here you're leaking! There is a "take" variant that will own `name` and free it, this one will `strdup()` and leak :-)

File: src/lib/common/sol-platform-impl-contiki.c (hunk 3 above)
Comment (lpereira) [inline src/lib/common/sol-platform-impl-contiki.c]: If it's unavailable, is it possible to generate one based on, say, the MAC address, or the machine ID, so it's always available, making it easier to obtain it (no error checking from the client-side, etc).

Comment (barbieri) [inline src/lib/common/sol-platform-impl-contiki.c]: we're not sure the OS have such concept. If they do not, then what we can do is a "fallback" implementation that only stores/reads from a global variable, so other soletta apps/subsys would work

File: src/lib/common/sol-platform-impl-linux-micro.c (hunk 5 above)
Comment (cabelitos) [inline src/lib/common/sol-platform-impl-linux-micro.c]: Ok

File: src/lib/common/sol-platform-impl-linux-micro.c (hunk 4 above)
Comment (cabelitos) [inline src/lib/common/sol-platform-impl-linux-micro.c]: Ok

File: src/lib/common/include/sol-platform.h (hunk 1 above)
Comment (cabelitos) [inline src/lib/common/include/sol-platform.h]: Ok

File: src/lib/common/sol-platform-impl-systemd.c (hunk 7 above)
Comment (cabelitos) [inline src/lib/common/sol-platform-impl-systemd.c]: I agree if you, I will change that!

File: src/lib/common/include/sol-platform.h (hunk 2 above)
Comment (cabelitos) [inline src/lib/common/include/sol-platform.h]: I think I will hide the complexity and create a static buffer.

File: src/modules/flow/platform/platform.c (hunk 8 above)
Comment (cabelitos) [inline src/modules/flow/platform/platform.c]: My bad. :(

**Comment by cabelitos** [comment]
Closing in favor of #1139

**Pull request closed without merging**